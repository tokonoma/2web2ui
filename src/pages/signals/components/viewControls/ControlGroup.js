import React, { Component } from 'react';
import PropTypes from 'prop-types';
import classNames from 'classnames';
import { Button } from 'src/components/matchbox';
import { useHibana } from 'src/context/HibanaContext';
import useHibanaOverride from 'src/hooks/useHibanaOverride';
import _ from 'lodash';
import OGStyles from './ControlGroup.module.scss';
import hibanaStyles from './ControlGroupHibana.module.scss';

class ControlGroup extends Component {
  state = {
    selected: '',
  };

  componentDidMount() {
    const { initialSelected } = this.props;

    if (initialSelected) {
      this.setState({ selected: initialSelected });
    }
  }

  handleChange = selected => {
    const { onChange } = this.props;

    this.setState({ selected });

    if (onChange) {
      onChange(selected);
    }
  };

  renderButtons = () => {
    const { options } = this.props;
    const { selected } = this.state;

    return _.keys(options).map(key => {
      return (
        <ControlButton
          key={key}
          selectedKey={key}
          isSelected={selected === key}
          selectedKey={key}
          onClick={() => this.handleChange(key)}
        >
          {options[key]}
        </ControlButton>
      );
    });
  };

  render() {
    return <Button.Group>{this.renderButtons()}</Button.Group>;
  }
}

function ControlButton(props) {
  const [state] = useHibana();
  const styles = useHibanaOverride(OGStyles, hibanaStyles);
  const { isHibanaEnabled } = state;
  const { children, selectedKey, isSelected, onClick } = props;

  let defaultStyleProps = {};
  let selectedStyleProps = {};

  if (isHibanaEnabled) {
    defaultStyleProps = {
      color: 'gray',
      outlineBorder: true,
    };

    selectedStyleProps = {
      color: 'gray',
      flat: false,
    };
  }

  return (
    <Button
      {...(isSelected ? selectedStyleProps : defaultStyleProps)}
      selectedkey={selectedKey}
      onClick={onClick}
      className={classNames(styles.Button, isSelected && styles.isSelected)}
      size="small"
    >
      {children}
    </Button>
  );
}

ControlGroup.propTypes = {
  options: PropTypes.object.isRequired,
  initialSelected: PropTypes.string,
  onChange: PropTypes.func,
};

export default ControlGroup;
