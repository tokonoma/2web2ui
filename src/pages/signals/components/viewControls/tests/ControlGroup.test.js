import { shallow } from 'enzyme';
import React from 'react';
import ControlGroup from '../ControlGroup';

jest.mock('src/hooks/useHibanaOverride', () => styles => styles);

describe('Signals View ControlGroup Component', () => {
  let wrapper;
  let props;

  beforeEach(() => {
    props = {
      options: {
        foo: 'fOo',
        bar: <test>BaR</test>,
      },
      initialSelected: 'bar',
      onChange: jest.fn(),
    };
    wrapper = shallow(<ControlGroup {...props} />);
  });

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });

  it('handles a select correctly', () => {
    wrapper
      .find('ControlButton')
      .first()
      .simulate('click');
    expect(wrapper.find('ControlButton').first()).toMatchSnapshot();
    expect(props.onChange).toHaveBeenCalledWith('foo');
  });

  it('does not call onChange if not provided a callback', () => {
    wrapper.setProps({ onChange: null });
    wrapper
      .find('ControlButton')
      .first()
      .simulate('click');
    expect(props.onChange).not.toHaveBeenCalled();
  });

  it('does not set selected initially if not provided', () => {
    wrapper = shallow(<ControlGroup options={props.options} />);
    expect(wrapper).toMatchSnapshot();
  });
});
