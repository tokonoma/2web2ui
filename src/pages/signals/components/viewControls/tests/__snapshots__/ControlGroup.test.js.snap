// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Signals View ControlGroup Component does not set selected initially if not provided 1`] = `
<Button.Group>
  <ControlButton
    isSelected={false}
    key="foo"
    onClick={[Function]}
    selectedKey="foo"
  >
    fOo
  </ControlButton>
  <ControlButton
    isSelected={false}
    key="bar"
    onClick={[Function]}
    selectedKey="bar"
  >
    <test>
      BaR
    </test>
  </ControlButton>
</Button.Group>
`;

exports[`Signals View ControlGroup Component handles a select correctly 1`] = `
<ControlButton
  isSelected={true}
  key="foo"
  onClick={[Function]}
  selectedKey="foo"
>
  fOo
</ControlButton>
`;

exports[`Signals View ControlGroup Component renders correctly 1`] = `
<Button.Group>
  <ControlButton
    isSelected={false}
    key="foo"
    onClick={[Function]}
    selectedKey="foo"
  >
    fOo
  </ControlButton>
  <ControlButton
    isSelected={true}
    key="bar"
    onClick={[Function]}
    selectedKey="bar"
  >
    <test>
      BaR
    </test>
  </ControlButton>
</Button.Group>
`;
